### ISSUE TITLE
[Q] Is it possible to plot metric where epochs is on the x-axis in mmDetection?

### ISSUE THREAD

**Issue by David-Biggs** [wandb/client#3910]
Thanks for the awesome tool!!

I am training multiple models each with a different training dataset size. I want to evaluate each model's performance . At the moment I seem only to be able to plot step (iteration) on the x-axis. This is not ideal as the smaller datasets would train for fewer steps, see the screenshot below. 

(This is due to the following formula `num_itterations = (num_epochs*num_imgs_in_dataset)/batch_size`. With the num_epochs and batch size remaining constant while the number of imgs in the dataset decreases, the number of iterations therefore decreasea)

I am using mmDetection and have not found a way to log the metrics per epoch and plot them in the Wandb UI. 

Any help would be much appreciated.

![image](https://user-images.githubusercontent.com/64862726/177936845-cf4c4ce0-6934-401f-a9ec-571a42ed623e.png)

**Comment by nate-wandb** [comment]
Hi @David-Biggs,
Would you prefer to only log your metric every epoch or are you looking to change the way it is visualized in the UI?

If you would like to only log once per epoch, could you share your training script and I can help you with this?

The other option would be to change the scale in the UI using the expressions tab. For instance you could do something like the following:
<img width="1366" alt="Screen Shot 2022-07-11 at 6 32 56 AM" src="https://user-images.githubusercontent.com/97066933/178266267-7ecd871c-b591-4f48-84cf-900d982f3fa2.png">

This could be variable based on different steps per epoch if you saved `steps_per_epoch` in the config using an expression like this: `_step/${config:steps_per_epoch}`. Even if you didn't save this config setting when you created the run you have the option to update this using the [public API](https://docs.wandb.ai/guides/track/public-api-guide#update-config-for-an-existing-run).

Thank you,
Nate

**Comment by David-Biggs** [comment]
Hi @nate-wandb , thanks for your reply.

I am hoping to log my metrics every epoch. I have runs which train for the same number of epochs but each run has a different number of steps per epoch. I am using sweeps and training in MMDetection. 

Here is my training script:
`def train():`
 `----with wandb.init() as run:`
        `--------wandbconfig = wandb.config`
        `--------cfg.optimizer.lr = wandbconfig.learning_rate`
        `--------cfg.optimizer.momentum = wandbconfig.momentum`
        `--------cfg.data.samples_per_gpu = wandbconfig.batch_size`
`--------model = build_detector(cfg.model)`
`--------train_detector(model, datasets, cfg, distributed=False, validate=True)`

I call `train()` within the agent `wandb.agent(sweep_id, train)`

Here is a screen shot of the output while training. 
![image](https://user-images.githubusercontent.com/64862726/178416179-64983211-165c-480d-9d21-3b3975983de4.png)

I have added  wandb to hooks. 

![image](https://user-images.githubusercontent.com/64862726/178416337-dc4737e5-d5f4-4466-89c5-251b2b6e6832.png)

I can change the print-out interval using this `cfg.log_config.interval = 10`

Thanks again for your help.

Regards,
David

**Comment by David-Biggs** [comment]
Hi @nate-wandb ,

Just following on this.

Many thanks,
David

**Comment by David-Biggs** [comment]
Hi Thanos,

Thanks for following up.

I've had a look the documentation. I see that EpochBasedRunner (`by_epcoh`) is by default set to True. 
I have set `with_step = False` and `by_epoch = True `. However, the metrics are still being plotted with Steps on the x-axis and not epochs. Below is a screenshot of the training and val losses. All three models are trained for the same number of epochs. 

![image](https://user-images.githubusercontent.com/64862726/180728313-f8b4d503-4620-4d69-a526-57daef991941.png)

Here is my WandbLoggerHook:

![image](https://user-images.githubusercontent.com/64862726/180729551-88e292a5-c227-40c0-bab9-55b23358d28b.png)

**Comment by thanos-wandb** [comment]
Hi @David-Biggs thank you for providing further information, it seems these options didn't do the trick. There is a way to define a custom step and log this as follows:

`# Define the custom x axis metric
wandb.define_metric("custom_step")

# Define which metrics to plot against that x-axis
wandb.define_metric("validation/loss", step_metric='custom_step')
for i in range(10):
  log_dict = {
      "train/loss": 1/(i+1),
      "custom_step": i**2,
      "validation/loss": 1/(i+1)   
  }
  wandb.log(log_dict)`

Or another way would be to do `wandb.log({'loss': loss}, step=step)`

Would any of these workarounds work for you?

**Comment by thanos-wandb** [comment]
Hi @David-Biggs I wanted to follow up on the issue you reported. Did any of the above suggestions resolve this? The other alternative would be that we file a feature request to MMCV.

**Comment by David-Biggs** [comment]
HI @thanos-wandb ,

Thanks for following up. 

I have not resolved the issue yet. I managed to find a messy workaround for now. I exported the data to a csv, and then plotted the mAP values in python. However, this does not resolve the issue for plotting the losses. They still appear per step. 

Your last comment would need to appear somewhere within MMCV's training loop but I can't seem to find where to put it...

How would I go about filing a feature request to MMCV?

**Comment by thanos-wandb** [comment]
Thanks @David-Biggs for letting me know, I understand that this workaround isn't ideal. Indeed the code changes need to happen on MMCV side, but we will submit the pull request for you. I will keep you posted here on any updates.

**Comment by thanos-wandb** [comment]
Hi @David-Biggs following up on this, I have made a feature request to the engineering team for creating a PR to MMCV. In the meantime, it would help us to know if you tried the above workaround using `define_metric`. You could do something as follows:

```
wandb.define_metric("epoch")
wandb.define_metric("validation/loss", step_metric='epoch')
for i in range(10):
  log_dict = {
      "epoch": the_epoch_var_from_MMCV,
      "validation/loss": 1/(i+1)   
  }
  wandb.log(log_dict)
```

Would this work for you?
